### PR TITLE
crud: improvments

### LINKED ISSUE #271: Follow-ups for CRUD API
1. Implement an ability to decode custom types via `Result`/`ResultMany`.
2. Add examples.
3. Add some benchmarks/make request types immutable.

CRUD API has been provided by PR #263.

### PR DESCRIPTION
**Pull request by oleg-jukovec**
The patchset makes a few changes:

1. It removes `NewEncoder`/`NewDecoder` from the public API.
2. It change a default tuple alias from `[]interface{}` to `interface{}`. This is necessary to use a custom types with custom encoders as tuples.
3. It adds examples with a request/response processing.
4. It removes dependency from `github.com/markphelps/optional`. The package gives a big overhead in benchmarks.
5. It makes requests immutable. It is more suitable for Go & GC. See also: https://github.com/tarantool/go-tarantool/issues/238
6. I removed special `Result*` types. They don't make much sense for `crud` since the response format is the same.
7. A user can specify a custom row type for `Result`.

I didn't forget about (remove if it is not applicable):

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

Related issues:

Closes #271

### CODE REVIEW

File: crud/options.go [hunk 1]
```diff
@@ -25,86 +19,95 @@ const (
...
+	value uint
+	exist bool
 }
 
-// NewOptUint creates an optional uint from value.
-func NewOptUint(value uint) OptUint {
-	return OptUint{optional.NewUint(value)}
+// MakeOptUint creates an optional uint from value.
+func MakeOptUint(value uint) OptUint {
+	return OptUint{
```
Comment (DifferentialOrange) [inline crud/options.go:30]: Are NewUint more expensive?

File: crud/count.go [hunk 2]
```diff
@@ -88,27 +88,27 @@ func NewCountRequest(space string) *CountRequest {
 
 // Conditions sets the conditions for the CountRequest request.
 // Note: default value is nil.
-func (req *CountRequest) Conditions(conditions []Condition) *CountRequest {
+func (req CountRequest) Conditions(conditions []Condition) CountRequest {
```
Comment (DifferentialOrange) [inline crud/count.go:91]: Does such setters allowed in go?

**Review by DifferentialOrange — APPROVED** [review]
Seems fine

File: crud/options.go (hunk 1 above)
Comment (oleg-jukovec) [reply crud/options.go:30]: This is not a very strict rule, but usually:
`NewFoo` return a pointer to a Foo object.
`MakeFoo` return an object of Foo.

File: crud/count.go (hunk 2 above)
Comment (oleg-jukovec) [reply crud/count.go:91]: Yea, there is a difference. A usage pattern should be:
```Go
mys2 = mys2.A(4)
```
You could see the pattern wih `SomeRequest` types in tests and examples.

> And from the first glance, it seems that it is some S3 with different handles.

Yes, there will be a difference with the main package, but it is highlighted by `Make` naming. I plan to make requests in the main package immutable too in `v2`.

> The breaking change is a separate question -- it seems that we rely on nobody using crud from master now

I think before a release we are free to change the behavior.

Comment (DifferentialOrange) [reply crud/count.go:91]: > I think before a release we are free to change the behavior.

Does specifying `https://github.com/tarantool/go-tarantool` fetches last tag or current master?

Comment (oleg-jukovec) [reply crud/count.go:91]: Yes, it will be a latest release version by default. You need to specify a latest commit manually for `master`.

**Review by DifferentialOrange — APPROVED** [review]

**Review by AnaNek — APPROVED** [review]

**Pull request merged**